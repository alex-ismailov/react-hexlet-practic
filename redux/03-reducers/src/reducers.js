import _ from 'lodash';
import { combineReducers } from 'redux';

const comments = (state = {}, action) => {
  // BEGIN (write your solution here)
  switch (action.type) {
    case 'TASK_COMMENT_ADD': {
      const { id, taskId } = action.payload.comment;
      return { ...state, [id]: { id, taskId } };
    }
    case 'TASK_COMMENT_REMOVE': {
      const { id } = action.payload;
      return _.omit(state, id);
    }
    case 'TASK_REMOVE': {
      const { id } = action.payload;
      const idKeys = Object.keys(state)
        .filter((key) => state[key].taskId === id);

      return _.omit(state, idKeys);
    }
    default:
      return state;
  }
  // END
};

const tasks = (state = {}, action) => {
  // BEGIN (write your solution here)
  switch (action.type) {
    case 'TASK_ADD': {
      const { id } = action.payload.task;
      return { ...state, [id]: { id } };
    }
    case 'TASK_REMOVE': {
      const { id } = action.payload;
      return _.omit(state, id);
    }
    default:
      return state;
  }
  // END
};

export default combineReducers({
  comments,
  tasks,
});
