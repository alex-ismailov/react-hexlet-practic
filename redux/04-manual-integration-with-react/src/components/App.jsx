/* eslint-disable react/static-property-placement */
import React from 'react';

// BEGIN (write your solution here)
export default class App extends React.Component {
  static defaultProps = {
    text: '',
  };

  handleInputText = (e) => {
    e.preventDefault();
    const { dispatch, updateText } = this.props;
    const { target: { value } } = e;
    dispatch(updateText(value));
  };

  handleResetText = (e) => {
    e.preventDefault();
    const { dispatch, resetText } = this.props;
    dispatch(resetText());
  };

  render() {
    const { text } = this.props;
    
    return (
      <div>
        <form>
          <input onChange={this.handleInputText} type="text" value={text} />
          <button onClick={this.handleResetText} type="button">Reset</button>
        </form>
        {text && <div>{text}</div>}
      </div>
    );
  }
}
// END
