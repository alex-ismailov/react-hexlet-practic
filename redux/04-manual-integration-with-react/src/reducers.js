import { combineReducers } from 'redux';

const text = (state = '', action) => {
  // BEGIN (write your solution here)
  switch (action.type) {
    case 'TEXT_UPDATE': {
      const { text } = action.payload;
      return text;
    }
    case 'TEXT_RESET': {
      return '';
    }
    default:
      return state;
  }
  // END
};

export default combineReducers({
  text,
});
