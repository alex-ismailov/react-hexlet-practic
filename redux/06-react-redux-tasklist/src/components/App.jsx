import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { updateNewTaskText, addTask, removeTask } from '../actions/index';

// BEGIN (write your solution here)
// const mapStateToProps = (state) => {
//   const props = {
//     text: state.text,
//     tasks: state.tasks,
//   };
//   return props;
// };

const mapStateToProps = ({ text, tasks }) => {
  const props = { text, tasks };
  return props;
};

class App extends React.Component {
  handleInputTaskText = (e) => {
    const { dispatch } = this.props;
    const { target: { value } } = e;
    dispatch(updateNewTaskText(value));
  };

  handleAddTask = (e) => {
    e.preventDefault();
    const { dispatch, text } = this.props;
    const newTask = { id: _.uniqueId(), text };
    dispatch(addTask(newTask));
  };

  handleRemoveTask = (id) => () => {
    const { dispatch } = this.props;
    dispatch(removeTask(id));
  };

  renderTasks() {
    const { tasks } = this.props;
    if (tasks.length === 0) {
      return null;
    }

    const taskItems = tasks.map((task) => (
      <li key={task.id} className="list-group-item d-flex">
        <span className="mr-auto">{task.text}</span>
        <button onClick={this.handleRemoveTask(task.id)} type="button" className="close">
          <span>&times;</span>
        </button>
      </li>
    ));

    return (
      <div className="mt-3">
        <ul className="list-group">{taskItems}</ul>
      </div>
    );
  }

  render() {
    const { text } = this.props;

    return (
      <div className="col-5">
        <form onSubmit={this.handleAddTask} action="" className="form-inline">
          <div className="form-group mx-sm-3">
            <input onChange={this.handleInputTaskText} type="text" required value={text} />
          </div>
          <button type="submit" className="btn btn-primary btn-sm">Add</button>
        </form>
        {this.renderTasks()}
      </div>
    );
  }
}

export default connect(mapStateToProps)(App);
// END
