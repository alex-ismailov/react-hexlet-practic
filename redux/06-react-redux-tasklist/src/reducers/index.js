
import { combineReducers } from 'redux';

const text = (state = '', action) => {
  switch (action.type) {
    case 'TEXT_UPDATE': {
      return action.payload.text;
    }
    case 'TASK_ADD': {
      return '';
    }
    default:
      return state;
  }
};

// BEGIN (write your solution here)
const tasks = (state = [], action) => {
  switch (action.type) {
    case 'TASK_ADD': {

      return;
    }
    case 'TASK_REMOVE': {

      return;
    }
    default:
      return state;
  }
};
// END

export default combineReducers({
  text,
  tasks,
});
