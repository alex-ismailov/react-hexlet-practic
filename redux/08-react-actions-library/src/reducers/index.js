import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions/index.js';

// BEGIN (write your solution here)
const {
  updateNewTaskText,
  addTask,
  removeTask,
} = actions;

const defaultState = {
  text: '',
  tasks: [],
};

const taskText = handleActions(
  {
    [updateNewTaskText]: (state, { payload: text }) => text,
    [addTask]: () => '',
  },
  defaultState.text,
);

const tasks = handleActions(
  {
    [addTask]: (state, { payload: task }) => [task, ...state],
    [removeTask]: (state, { payload: id }) => state.filter((task) => task.id !== id),
  },
  defaultState.tasks,
);
// END

export default combineReducers({
  taskText,
  tasks,
});
