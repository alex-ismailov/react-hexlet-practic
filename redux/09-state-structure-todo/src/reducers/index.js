import _ from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions/index.js';

const tasks = handleActions({
  [actions.addTask](state, { payload: { task } }) {
    const { byId, allIds } = state;
    return {
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },
  [actions.removeTask](state, { payload: { id } }) {
    const { byId, allIds } = state;
    return {
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
    };
  },
  // BEGIN (write your solution here)
  [actions.toggleTaskState]: (state, { payload: { id } }) => {
    const newTaskState = state.byId[id].state === 'active'
      ? 'finished'
      : 'active';

    const newState = { ...state };
    _.set(newState, ['byId', id, 'state'], newTaskState);

    return newState;
  },
  // END
}, { byId: {}, allIds: [] });

const text = handleActions({
  [actions.addTask]() {
    return '';
  },
  [actions.updateNewTaskText](_state, { payload }) {
    return payload.text;
  },
}, '');

export default combineReducers({
  tasks,
  text,
});
