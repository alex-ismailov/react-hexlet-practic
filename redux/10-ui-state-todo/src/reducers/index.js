import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions/index.js';

const tasks = handleActions({
  [actions.addTask](state, { payload: { task } }) {
    const { byId, allIds } = state;
    return {
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },
}, { byId: {}, allIds: [] });

const tasksUIState = handleActions({
  // BEGIN (write your solution here)
  
  // END
}, {});

const text = handleActions({
  [actions.addTask]() {
    return '';
  },
  [actions.updateNewTaskText](state, { payload }) {
    return payload.text;
  },
}, '');

export default combineReducers({
  tasks,
  tasksUIState,
  text,
});
