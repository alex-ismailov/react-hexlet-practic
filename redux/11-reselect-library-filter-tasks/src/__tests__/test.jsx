import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducers from '../reducers/index.js';
import App from '../components/App.jsx';

Enzyme.configure({ adapter: new Adapter() });

test('Store', () => {
  const store = createStore(reducers);

  const vdom = (
    <Provider store={store}>
      <App />
    </Provider>
  );
  const wrapper = mount(vdom);
  expect(wrapper.render()).toMatchSnapshot();

  const newTaskInput = wrapper.find('input[type="text"]');
  const newTaskSubmit = wrapper.find('input[type="submit"]');

  newTaskInput.simulate('change', { target: { value: 'na-na' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'another task' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  const link = wrapper.find('[data-test="task-toggle-state"]').first();
  link.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const activeFilterButton = wrapper.find('[data-test="task-filter-active"]');
  const finishedFilterButton = wrapper.find('[data-test="task-filter-finished"]');

  activeFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  finishedFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const allFilterButton = wrapper.find('[data-test="task-filter-all"]');

  allFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const links = wrapper.find('[data-test="task-remove"]');
  links.last().simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
