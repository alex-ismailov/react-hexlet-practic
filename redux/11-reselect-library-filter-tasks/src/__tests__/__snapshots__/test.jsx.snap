// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
</div>
`;

exports[`Store 2`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value="na-na"
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
</div>
`;

exports[`Store 3`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              na-na
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 4`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value="another task"
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              na-na
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 5`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              another task
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              na-na
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 6`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              <s>
                another task
              </s>
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              na-na
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 7`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-all"
      type="button"
    >
      All Tasks
    </button>
    Active Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              na-na
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 8`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-all"
      type="button"
    >
      All Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    Finished Tasks
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              <s>
                another task
              </s>
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 9`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              <s>
                another task
              </s>
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              na-na
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 10`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <input
      class="btn btn-primary btn-sm"
      type="submit"
      value="Add"
    />
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <button
              class="btn btn-link"
              data-test="task-toggle-state"
              type="button"
            >
              <s>
                another task
              </s>
            </button>
          </span>
          <button
            class="close"
            data-test="task-remove"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
